MekHQ/src/mekhq/campaign/Campaign.java

```diff
@@ -3730,6 +3739,45 @@ && getLocation().getJumpPath().getLastSystem().getId().equals(contract.getSystem
         }
     }
 
+    /**
+     * Processes an abandoned convoy. The player is presented with a defeat dialog,
+     * and if they have independent command rights, it checks each player template force in the scenario
+     * for being a convoy force. If it is a convoy force, its units are treated as abandoned units.
+     * Each crew member of these units is set as either KIA or POW based on a die roll.
+     * It finally removes each unit from the campaign.
+     *
+     * @param contract The current {@link AtBContract}.
+     * @param scenario The relevant {@link AtBDynamicScenario}.
+     */
+    private void processAbandonedConvoy(AtBContract contract, AtBDynamicScenario scenario) {
+        convoyFinalMessageDialog(this, contract.getEmployerFaction());
+
+        if (contract.getCommandRights().isIndependent()) {
+            for (Integer forceId : scenario.getPlayerTemplateForceIDs()) {
+                Force force = getForce(forceId);
+
+                if (force != null && force.isConvoyForce()) {
+                    for (UUID unitID : new ArrayList<>(force.getUnits())) {
+                        Unit unit = getUnit(unitID);
+                        if (unit != null) {
+                            List<Person> crew = new ArrayList<>(unit.getCrew());
+                            for (Person crewMember : crew) {
+                                PersonnelStatus status = KIA;
+                                // We're using the CamOps rules for infantry survival here and
+                                // assuming anyone who isn't dead has been captured.
+                                if (Compute.d6(2) > 7) {
+                                    status = PersonnelStatus.POW;
+                                }
+                                crewMember.changeStatus(this, currentDay, status);
+                            }
+                        }
+                        removeUnit(unitID);
+                    }
+                }
```

Comment:
This is a suggestion on how to use modern java with streams for data manipulation and reduce the number of branching operation in the CPU.

The following code was not tested and possibly has bug (like the simultaneos removal of units while iterate over its list, probably need to keep new ArrayList in there (commented out, and probably slightly wrong.

```java
    /**
     * Processes an abandoned convoy. The player is presented with a defeat dialog,
     * and if they have independent command rights, it checks each player template force in the scenario
     * for being a convoy force. If it is a convoy force, its units are treated as abandoned units.
     * Each crew member of these units is set as either KIA or POW based on a die roll.
     * It finally removes each unit from the campaign.
     *
     * @param contract The current {@link AtBContract}.
     * @param scenario The relevant {@link AtBDynamicScenario}.
     */
    private void processAbandonedConvoy(AtBContract contract, AtBDynamicScenario scenario) {
        convoyFinalMessageDialog(this, contract.getEmployerFaction());

        if (!contract.getCommandRights().isIndependent()) {
            return;
        }

        scenario.getPlayerTemplateForceIDs().stream()
            .map(this::getForce).filter(Objects::nonNull)
            .filter(Force::isConvoyForce)
            .map(Force::getUnits)
            // .flatMap(new::ArrayList) ??? maybe this is necessary here? I cant test this case now
            .map(this::decideCrewFate)
            .forEach(this::removeUnit);
    }
    
    /**
     * decideCrewFate
     * <p>
     *     Decide the fate of the crew of a unit that was abandoned in a time of need
     * </p>
     * @param unitID
     * @return
     *
     */
    private UUID decideCrewFate(UUID unitID) {
        Optional.of(getUnit(unitID)).ifPresent(unit -> unit.getCrew().forEach(this::decideCrewmemberFate));
        return unitID;
    }

    /**
     * decidePersonFate
     * <p>
     *      We're using the CamOps rules for infantry survival here and
     *      assuming anyone who isn't dead has been captured.
     * </p>
     * @param person The person to decide the fate of.
     */
    private void decideCrewmemberFate(Person person) {
        PersonnelStatus status = KIA;
        if (Compute.d6(2) > 7) {
            status = PersonnelStatus.POW;
        }
        person.changeStatus(this, currentDay, status);
    }
```